### PR TITLE
Add an overload to Object.freeze that preserves literal types

### PR DESCRIPTION
**Pull request by weswigham**
This way when you write something like
```js
const LogType = Object.freeze({
	error: "error", // message, c style arguments
	warn: "warn", // message, c style arguments
	info: "info", // message, c style arguments
	log: "log", // message, c style arguments
	debug: "debug", // message, c style arguments

	trace: "trace", // no arguments

	group: "group", // [label]
	groupCollapsed: "groupCollapsed", // [label]
	groupEnd: "groupEnd", // [label]

	profile: "profile", // [profileName]
	profileEnd: "profileEnd", // [profileName]

	time: "time", // name, time as [seconds, nanoseconds]

	clear: "clear", // no arguments
	status: "status" // message, arguments
});
```
(snippet from webpack) you get something with literally-typed members as a result, if possible. The intuition seems to be that it should just work like that already, but it does not currently.

### CODE REVIEW

File: src/lib/es5.d.ts [hunk 1]
```diff
@@ -204,6 +204,12 @@ interface ObjectConstructor {
      */
     freeze<T extends Function>(f: T): T;
 
+    /**
+     * Prevents the modification of existing property attributes and values, and prevents the addition of new properties.
+     * @param o Object on which to lock the attributes.
+     */
+    freeze<T extends {[idx: string]: U | null | undefined | object}, U extends string | number | symbol>(o: T): Readonly<T>;
```
Comment (falsandtru) [inline src/lib/es5.d.ts]: Should has bigint?

Comment (weswigham) [reply src/lib/es5.d.ts]: Ah, yes.

**Comment by weswigham** [comment]
@typescript-bot test this

**Comment by weswigham** [comment]
@typescript-bot user test this

**Comment by sandersn** [comment]
@typescript-bot user test this inline

**Comment by sandersn** [comment]
@weswigham this seems like a plausible change even though it's old. It didn't create any new errors in the user tests. Think it's worth trying out?

**Comment by weswigham** [comment]
Yeah, I think it's still a reasonable change.

**Review by sandersn — APPROVED** [review]

**Comment by Andarist** [comment]
Out of curiosity... if I hover over the call using this signature I might see something like this:
```ts
function freeze<{
    a: 1;
}, string | number | bigint | symbol>(o: {
    a: 1;
}): Readonly<{
    a: 1;
}>
```

In here, the `U` was not "computed" but instead its constraint is displayed. What kind of TS behavior/trait is at play here? On what foundation/principles this PR is built? 🤔

**Comment by weswigham** [comment]
We assign positions literal types when they're contextually typed by a generic with literal bases in the constraint.

**Pull request merged**